### DISCUSSION TITLE
This is safe using rsync for combining two dirs to another one?

### DISCUSSION THREAD

**Discussion by Azq2** [RsyncProject/rsync#344] [Q&A] [answered]
For e.g. I want to merge all files from `overlay/` and `main/` to one dir `destination/`
And files in `overlay/` must to have priority on files from `main/`

I use this command:
```
rsync -av --delete overlay/ main/ destination/
```

This works... but I'm not sure if that safe.

Is it guaranteed that files from the first source will always be in priority?
I see in the code this algorithm:

- get all files from all source dirs
- sort this files by fsort() (by name)
- remove all duplicates, use only first file with concrete name

But will the file from the first source always come first after fsort?
Or this is undefined behavior? 

P.S. I want to use single rsync command.

**Comment by Azq2** [comment]
Explanation of my case

```bash
# create test data
$ mkdir overlay/ main/ destination/
$ echo "file from overlay" > overlay/file
$ echo "file from main" > main/file

# rsync files
$ rsync -av --delete overlay/ main/ destination/
sending incremental file list
./
file

sent 168 bytes  received 38 bytes  412.00 bytes/sec
total size is 12  speedup is 0.06

# check result
$ cat destination/file
file from overlay
```

**Comment by Azq2** [comment]
Also I read this in wiki:

> In computer science, merge sort (also commonly spelled as mergesort) is an efficient, general-purpose, and comparison-based sorting algorithm. Most implementations produce a stable sort, **which means that the order of equal elements is the same in the input and output.**

Merge sort (fsort), used in rsync, is an stable sort?
I'm not specialist in algorithm and fsort code don't tell me anything about of stability :(

**Comment by WayneD** [comment, marked as answer]
The flist.c file has comments about this, but I didn't find it mentioned in the manpage.  Some of the comments:

> /* This file-struct sorting routine makes sure that any identical names in
> ** the file list stay in the same order as they were in the original list.
> ** This is particularly vital in inc_recurse mode where we expect a sort
> ** on the flist to match the exact order of a sort on the dir_flist. */

> /* If one is a dir and the other is not, we want to
> ** keep the dir because it might have contents in the
> ** list.  Otherwise keep the first one. */

So, with identically named files, rsync always chooses the first one encountered. If the second folder has a dir named the same as a file in the first folder, the dir is sent instead.

**Comment by Azq2** [comment]
Thanks for explanation. Rsync is awesome :)
